{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7733999",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import scanpy as sc\n",
    "import scanpy.external as sce\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import string\n",
    "import anndata\n",
    "data = '/home/carsten/alvira_bioinformatics/etx_rat_scrnaseq/data/CellRanger_output'\n",
    "sc.set_figure_params(dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c3c1e47",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = '/home/carsten/alvira_bioinformatics/etx_rat_scrnaseq/data/processed/single_cell_files'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0908677",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read(f'{data}/etx_rat_processed_cell_typed_raw.gz.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80dd100b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.groupby(['Timepoint','Treatment'])['Cell Subtype'].value_counts(normalize=True).to_csv('value_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24492579",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea6f0022",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[~adata.obs['Cell Subtype'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f148a928",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata, target_sum = 1e4)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1932cded",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata,\n",
    "          color= 'Gas7')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bedab7e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata,\n",
    "          color= 'Scn7a')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b73e6ef5",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding(adata[adata.obs['lineage']=='endothelial'],\n",
    "               basis = 'X_umap_endothelial',\n",
    "               color = 'Cell Subtype')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "611a2243",
   "metadata": {},
   "outputs": [],
   "source": [
    "mes= adata[adata.obs['lineage']=='mesenchymal']\n",
    "mes_d0 = mes[mes.obs['Timepoint']=='E20']\n",
    "sc.pp.pca(mes_d0)\n",
    "sc.pp.neighbors(mes_d0)\n",
    "sc.tl.umap(mes_d0)\n",
    "sc.pl.umap(mes_d0, color = 'Cell Subtype')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3412feb",
   "metadata": {},
   "outputs": [],
   "source": [
    "mes= adata[adata.obs['lineage']=='mesenchymal']\n",
    "mes_d0 = mes[mes.obs['Timepoint']=='D0']\n",
    "sc.pp.normalize_total(mes_d0, target_sum = 1e4)\n",
    "sc.pp.log1p(mes_d0)\n",
    "sc.pp.pca(mes_d0)\n",
    "sc.pp.neighbors(mes_d0)\n",
    "sc.tl.umap(mes_d0)\n",
    "sc.pl.umap(mes_d0, color = 'Cell Subtype')\n",
    "sc.pl.umap(mes_d0, color = 'Mki67')\n",
    "sc.pl.umap(mes_d0, color = 'Treatment')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10c891e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "mes= adata[adata.obs['lineage']=='mesenchymal']\n",
    "mes_d0 = mes[mes.obs['Timepoint']=='D0']\n",
    "sc.pp.pca(mes_d0)\n",
    "sc.pp.neighbors(mes_d0)\n",
    "sc.tl.umap(mes_d0)\n",
    "sc.pl.umap(mes_d0, color = 'Cell Subtype')\n",
    "sc.pl.umap(mes_d0, color = 'Mki67')\n",
    "sc.pl.umap(mes_d0, color = 'Treatment')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd4c7239",
   "metadata": {},
   "outputs": [],
   "source": [
    "mes= adata[adata.obs['lineage']=='mesenchymal']\n",
    "mes_d0 = mes[mes.obs['Timepoint']=='D0']\n",
    "sce.pp.bbknn(\n",
    "    mes_d0,\n",
    "    batch_key=\"Run\",\n",
    "    n_pcs=10,\n",
    "    neighbors_within_batch=10,\n",
    "    set_op_mix_ratio=0.9,\n",
    ")\n",
    "sc.tl.umap(mes_d0)\n",
    "sc.pl.umap(mes_d0, color = 'Cell Subtype')\n",
    "sc.pl.umap(mes_d0, color = 'Mki67')\n",
    "sc.pl.umap(mes_d0, color = 'Treatment')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b52389e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8832c4d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(mes_d0, color = 'Pdgfra')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c3facaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "mes= adata[adata.obs['lineage']=='mesenchymal']\n",
    "mes_d0 = mes[mes.obs['Timepoint']=='E20']\n",
    "sc.pp.neighbors(mes_d0)\n",
    "sc.tl.umap(mes_d0)\n",
    "sc.pl.umap(mes_d0, color = 'Cell Subtype')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f5f5c71",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(mes_d0, color = ['Limch1','Col14a1','Col13a1', 'Macf1','Dcn','Mgp','Tnc','Rbp1', 'Pdgfra'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2609142",
   "metadata": {},
   "outputs": [],
   "source": [
    "mes= adata[adata.obs['lineage']=='mesenchymal']\n",
    "mes_d14 = mes[mes.obs['Timepoint']=='D14']\n",
    "sc.pp.neighbors(mes_d14)\n",
    "sc.tl.umap(mes_d14)\n",
    "sc.pl.umap(mes_d14, color = 'Cell Subtype')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb0d9506",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding(end, color = 'leiden_endothelial', basis = 'X_umap_endothelial')\n",
    "big_v = end[end.obs['leiden_endothelial'].isin(['5','6','10'])]\n",
    "sc.tl.rank_genes_groups(big_v,groupby='leiden_endothelial',method='wilcoxon')\n",
    "sc.pl.rank_genes_groups_dotplot(big_v,groupby ='leiden_endothelial', n_genes=10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39fc78c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding(end, color = 'Igf2', basis = 'X_umap_endothelial')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8430485b",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_dict = {\n",
    "        \"mesenchymal\": [\n",
    "            \"Col3a1\",\n",
    "            \"G0s2\",\n",
    "            \"Limch1\",\n",
    "            \"Col13a1\",\n",
    "            \"Col14a1\",\n",
    "            \"Serpinf1\",\n",
    "            \"Pdgfra\",\n",
    "            \"Scara5\",\n",
    "            \"Acta2\",\n",
    "            \"Hhip\",\n",
    "            \"Fgf18\",\n",
    "            \"Wif1\",\n",
    "            \"Tgfbi\",\n",
    "            \"Tagln\",\n",
    "            \"Mustn1\",\n",
    "            \"Aard\",\n",
    "            \"Pdgfrb\",\n",
    "            \"Cox4i2\",\n",
    "            # 'Higd1b',\n",
    "            \"Myh11\",\n",
    "            'Msln',\n",
    "            \"Wt1\",\n",
    "            \"Lrrn4\",\n",
    "            \"Upk3b\",\n",
    "            \"Mki67\",\n",
    "            \"Acta1\",\n",
    "            # 'Crh',\n",
    "            'Lgr6',\n",
    "            'Aspn',\n",
    "            'Cnn1',\n",
    "            \"Epcam\",\n",
    "            \"Ptprc\",\n",
    "            \"Pecam1\",\n",
    "        ],\n",
    "        \"endothelial\": [\n",
    "            \"Gja5\",\n",
    "            \"Bmx\",\n",
    "            \"Fn1\",\n",
    "            \"Ctsh\",\n",
    "            # 'Kcne3',\n",
    "            \"Cdh13\",\n",
    "            # 'Car8',\n",
    "            \"Mmp16\",\n",
    "            \"Slc6a2\",\n",
    "            \"Thy1\",\n",
    "            \"Mmrn1\",\n",
    "            # 'Ccl21a',\n",
    "            \"Reln\",\n",
    "            \"Neil3\",\n",
    "            \"Mki67\",\n",
    "            \"Aurkb\",\n",
    "            \"Depp1\",\n",
    "            \"Ankrd37\",\n",
    "            \"Peg3\",\n",
    "            \"Mest\",\n",
    "            \"Hpgd\",\n",
    "            # 'Cd36',\n",
    "            # 'Car4',\n",
    "            \"Sirpa\",\n",
    "            \"Fibin\",\n",
    "            \"Col1a1\",\n",
    "            \"Epcam\",\n",
    "            \"Ptprc\",\n",
    "            \"Pecam1\",\n",
    "        ],\n",
    "        \"immune\": [\n",
    "            \"Cd68\",\n",
    "            \"Gal\",\n",
    "            \"Itgax\",\n",
    "            'Ca4',\n",
    "            \"C1qa\",\n",
    "            'Ccl3',\n",
    "            'Nkg7',\n",
    "            \"Plac8\",\n",
    "            \"Batf3\",\n",
    "            'Rora',\n",
    "            \"Itgae\",\n",
    "            # 'Cd209a',\n",
    "            'Ccl5',\n",
    "            \"Mreg\",\n",
    "            # 'Mcpt8',\n",
    "            \"Retnlg\",\n",
    "            \"Ms4a1\",\n",
    "            \"Gzma\",\n",
    "            \"Cd3e\",\n",
    "            \"Areg\",\n",
    "            \"Mki67\",\n",
    "            \"Col1a1\",\n",
    "            \"Epcam\",\n",
    "            \"Ptprc\",\n",
    "            \"Pecam1\",\n",
    "        ],\n",
    "        \"epithelial\": [\n",
    "            \"Scg5\",\n",
    "            \"Ascl1\",\n",
    "            # 'Lyz1',\n",
    "            \"Lyz2\",\n",
    "            \"Sftpc\",\n",
    "            \"Slc34a2\",\n",
    "            \"S100g\",\n",
    "            \"Sftpa1\",\n",
    "            \"Akap5\",\n",
    "            \"Hopx\",\n",
    "            # 'Col4a4',\n",
    "            \"Vegfa\",\n",
    "            \"Tmem212\",\n",
    "            \"Dynlrb2\",\n",
    "            \"Cdkn1c\",\n",
    "            \"Tppp3\",\n",
    "            # 'Scgb3a2',\n",
    "            # 'Cyp2f2',\n",
    "            \"Scgb1a1\",\n",
    "            # 'Reg3g',\n",
    "            \"Scgb3a1\",\n",
    "            \"Mki67\",\n",
    "            \"Col1a1\",\n",
    "            \"Epcam\",\n",
    "            \"Ptprc\",\n",
    "            \"Pecam1\",\n",
    "        ],\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ea69b56",
   "metadata": {},
   "outputs": [],
   "source": [
    "lineage ='immune'\n",
    "lin_adata = anndata.AnnData(\n",
    "            adata[adata.obs[\"lineage\"] == lineage].X,\n",
    "            obs=adata[adata.obs[\"lineage\"] == lineage].obs,\n",
    "            var=adata[adata.obs[\"lineage\"] == lineage].var,\n",
    "        )\n",
    "del lin_adata.var[\"highly_variable\"]\n",
    "sc.pp.highly_variable_genes(lin_adata, n_top_genes=2000, batch_key=\"Run\")\n",
    "sc.pp.pca(lin_adata, use_highly_variable=True)\n",
    "sce.pp.bbknn(\n",
    "            lin_adata,\n",
    "            batch_key=\"Run\",\n",
    "            n_pcs=10,\n",
    "            neighbors_within_batch=5,\n",
    "#     trim = 25,\n",
    "            set_op_mix_ratio=0.5,\n",
    "        )\n",
    "# sc.pp.neighbors(lin_adata)\n",
    "sc.tl.umap(lin_adata, a=2, b=1.5)\n",
    "sc.tl.leiden(lin_adata)\n",
    "for color in ['leiden','Cell Subtype','Run','Timepoint','Treatment',gene_dict[lineage]]:\n",
    "    if color == gene_dict[lineage]:\n",
    "        save = 'genes'\n",
    "    else:\n",
    "        save = color\n",
    "    sc.pl.umap(lin_adata, color = color, save=f'{lineage}_{save}.png')\n",
    "sc.tl.rank_genes_groups(lin_adata, f\"leiden\", method=\"wilcoxon\")\n",
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    lin_adata,\n",
    "    groupby=f\"leiden\",\n",
    "    n_genes=int(150 / len(lin_adata.obs[f\"leiden\"].unique())),\n",
    "    show=True,\n",
    "    save = f'{lineage}.png'\n",
    ")\n",
    "lin_adata.obs.groupby(['Timepoint','Treatment'])['Cell Subtype'].value_counts(normalize=True)*100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94a5606f",
   "metadata": {},
   "outputs": [],
   "source": [
    "lineage ='mesenchymal'\n",
    "lin_adata = anndata.AnnData(\n",
    "            adata[adata.obs[\"lineage\"] == lineage].X,\n",
    "            obs=adata[adata.obs[\"lineage\"] == lineage].obs,\n",
    "            var=adata[adata.obs[\"lineage\"] == lineage].var,\n",
    "        )\n",
    "del lin_adata.var[\"highly_variable\"]\n",
    "sc.pp.highly_variable_genes(lin_adata, n_top_genes=2000, batch_key=\"Run\")\n",
    "sc.pp.pca(lin_adata, use_highly_variable=True)\n",
    "sce.pp.bbknn(\n",
    "            lin_adata,\n",
    "            batch_key=\"Run\",\n",
    "            n_pcs=10,\n",
    "            neighbors_within_batch=10,\n",
    "    trim = 0,\n",
    "            set_op_mix_ratio=0.5,\n",
    "#     local_connectivity=1\n",
    "\n",
    "        )\n",
    "# sc.pp.neighbors(lin_adata)\n",
    "\n",
    "sc.tl.umap(lin_adata, a=2, b=1)\n",
    "sc.tl.leiden(lin_adata)\n",
    "\n",
    "for color in ['leiden','Cell Subtype','Run','Timepoint','Treatment',gene_dict[lineage]]:\n",
    "    if color == gene_dict[lineage]:\n",
    "        save = 'genes'\n",
    "    else:\n",
    "        save = color\n",
    "    sc.pl.umap(lin_adata, color = color, save=f'{lineage}_{save}.png')\n",
    "sc.tl.rank_genes_groups(lin_adata, f\"leiden\", method=\"wilcoxon\")\n",
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    lin_adata,\n",
    "    groupby=f\"leiden\",\n",
    "    n_genes=int(150 / len(lin_adata.obs[f\"leiden\"].unique())),\n",
    "    show=True,\n",
    "    save = f'{lineage}.png'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "924b739a",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(lin_adata.obs.groupby(['leiden','Treatment'])['Timepoint'].value_counts(normalize=True)*100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a537506",
   "metadata": {},
   "outputs": [],
   "source": [
    "asm_myof = lin_adata[lin_adata.obs['leiden'].isin(['3','4','1','5'])]\n",
    "sc.tl.rank_genes_groups(asm_myof, f\"leiden\", method=\"wilcoxon\")\n",
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    asm_myof,\n",
    "    groupby=f\"leiden\",\n",
    "    n_genes=10,\n",
    "    dendrogram=False,\n",
    "    show=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7dcbb87",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(lin_adata, color = ['Tnc','Actg2','Thbs1','Mgp','Ccn2','Cd34','Etv1','Wif1','Mmp2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d396c11",
   "metadata": {},
   "outputs": [],
   "source": [
    "D14 = lin_adata[lin_adata.obs['Timepoint']=='D14']\n",
    "D14 = anndata.AnnData(\n",
    "            D14[D14.obs[\"lineage\"] == lineage].X,\n",
    "            obs=D14[D14.obs[\"lineage\"] == lineage].obs,\n",
    "            var=D14[D14.obs[\"lineage\"] == lineage].var,\n",
    "        )\n",
    "del D14.var[\"highly_variable\"]\n",
    "sc.pp.highly_variable_genes(D14, n_top_genes=2000, batch_key=\"Run\")\n",
    "sc.pp.pca(D14, use_highly_variable=True)\n",
    "# sce.pp.bbknn(\n",
    "#             D14,\n",
    "#             batch_key=\"Run\",\n",
    "#             n_pcs=10,\n",
    "#             neighbors_within_batch=10,\n",
    "#     trim = 0,\n",
    "#             set_op_mix_ratio=0.5,\n",
    "# #     local_connectivity=1\n",
    "\n",
    "#         )\n",
    "sc.pp.neighbors(D14)\n",
    "\n",
    "sc.tl.umap(D14, a=2, b=1)\n",
    "sc.tl.leiden(D14)\n",
    "\n",
    "for color in ['leiden','Cell Subtype','Run','Timepoint','Treatment',gene_dict[lineage]]:\n",
    "    if color == gene_dict[lineage]:\n",
    "        save = 'genes'\n",
    "    else:\n",
    "        save = color\n",
    "    sc.pl.umap(D14, color = color, \n",
    "#                save=f'{lineage}_{save}.png'\n",
    "              )\n",
    "sc.tl.rank_genes_groups(D14, f\"leiden\", method=\"wilcoxon\")\n",
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    D14,\n",
    "    groupby=f\"leiden\",\n",
    "    n_genes=int(150 / len(D14.obs[f\"leiden\"].unique())),\n",
    "    show=True,\n",
    "#     save = f'{lineage}.png'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40ff82b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(D14, color =['Mmp2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd5f4d49",
   "metadata": {},
   "outputs": [],
   "source": [
    "lineage ='endothelial'\n",
    "lin_adata = anndata.AnnData(\n",
    "            adata[adata.obs[\"lineage\"] == lineage].X,\n",
    "            obs=adata[adata.obs[\"lineage\"] == lineage].obs,\n",
    "            var=adata[adata.obs[\"lineage\"] == lineage].var,\n",
    "        )\n",
    "del lin_adata.var[\"highly_variable\"]\n",
    "sc.pp.highly_variable_genes(lin_adata, n_top_genes=2000, batch_key=\"Run\")\n",
    "sc.pp.pca(lin_adata, use_highly_variable=True)\n",
    "sce.pp.bbknn(\n",
    "            lin_adata,\n",
    "            batch_key=\"Run\",\n",
    "            n_pcs=10,\n",
    "            neighbors_within_batch=5,\n",
    "    trim = 25,\n",
    "            set_op_mix_ratio=0.75,\n",
    "        )\n",
    "# sc.pp.neighbors(lin_adata)\n",
    "\n",
    "sc.tl.umap(lin_adata, a=2, b=0.8)\n",
    "sc.tl.leiden(lin_adata)\n",
    "for color in ['leiden','Cell Subtype','Run','Timepoint','Treatment',gene_dict[lineage]]:\n",
    "    if color == gene_dict[lineage]:\n",
    "        save = 'genes'\n",
    "    else:\n",
    "        save = color\n",
    "    sc.pl.umap(lin_adata, color = color, save=f'{lineage}_{save}.png')\n",
    "sc.tl.rank_genes_groups(lin_adata, f\"leiden\", method=\"wilcoxon\")\n",
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    lin_adata,\n",
    "    groupby=f\"leiden\",\n",
    "    n_genes=int(150 / len(lin_adata.obs[f\"leiden\"].unique())),\n",
    "    show=True,\n",
    "    save = f'{lineage}.png'\n",
    ")\n",
    "lin_adata.obs.groupby(['Timepoint','Treatment'])['Cell Subtype'].value_counts(normalize=True)*100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95b90ea5",
   "metadata": {},
   "outputs": [],
   "source": [
    "lineage ='epithelial'\n",
    "lin_adata = anndata.AnnData(\n",
    "            adata[adata.obs[\"lineage\"] == lineage].X,\n",
    "            obs=adata[adata.obs[\"lineage\"] == lineage].obs,\n",
    "            var=adata[adata.obs[\"lineage\"] == lineage].var,\n",
    "        )\n",
    "del lin_adata.var[\"highly_variable\"]\n",
    "sc.pp.highly_variable_genes(lin_adata, n_top_genes=2000, batch_key=\"Run\")\n",
    "sc.pp.pca(lin_adata, use_highly_variable=True)\n",
    "sce.pp.bbknn(\n",
    "            lin_adata,\n",
    "            batch_key=\"Run\",\n",
    "            n_pcs=10,\n",
    "            neighbors_within_batch=5,\n",
    "        trim = 25,\n",
    "            set_op_mix_ratio=0.75,\n",
    "        )\n",
    "# sc.pp.neighbors(lin_adata)\n",
    "sc.tl.umap(lin_adata, a=2, b=1.2)\n",
    "sc.tl.leiden(lin_adata)\n",
    "for color in ['leiden','Cell Subtype','Run','Timepoint','Treatment',gene_dict[lineage]]:\n",
    "    if color == gene_dict[lineage]:\n",
    "        save = 'genes'\n",
    "    else:\n",
    "        save = color\n",
    "    sc.pl.umap(lin_adata, color = color, save=f'{lineage}_{save}.png')\n",
    "sc.tl.rank_genes_groups(lin_adata, f\"leiden\", method=\"wilcoxon\")\n",
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    lin_adata,\n",
    "    groupby=f\"leiden\",\n",
    "    n_genes=int(150 / len(lin_adata.obs[f\"leiden\"].unique())),\n",
    "    show=True,\n",
    "    save = f'{lineage}.png'\n",
    ")\n",
    "lin_adata.obs.groupby(['Timepoint','Treatment'])['Cell Subtype'].value_counts(normalize=True)*100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95eec1c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_ls = []\n",
    "for k in gene_dict.keys():\n",
    "    gene_ls = gene_ls + gene_dict[k]\n",
    "gene_ls = gene_ls + [\n",
    "             'leiden',\n",
    "             'lineage',\n",
    "             'n_genes_by_counts',\n",
    "             'Run',\n",
    "             'Treatment',\n",
    "             'total_counts',\n",
    "             'Timepoint',\n",
    "             'doublet',\n",
    "             'doublet_score',\n",
    "             'pct_counts_mt',\n",
    "    'Cell Subtype'\n",
    "             ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b87f1117",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata2 = anndata.AnnData(\n",
    "            adata.X,\n",
    "            obs=adata.obs,\n",
    "            var=adata.var,\n",
    "        )\n",
    "del adata2.var[\"highly_variable\"]\n",
    "sc.pp.highly_variable_genes(adata2, n_top_genes=2000, batch_key=\"Run\")\n",
    "sc.pp.pca(adata2, use_highly_variable=True)\n",
    "# sc.pp.neighbors(adata2)\n",
    "sce.pp.bbknn(\n",
    "            adata2,\n",
    "            batch_key=\"Run\",\n",
    "            n_pcs=10,\n",
    "            neighbors_within_batch=5,\n",
    "    trim = 25,\n",
    "            set_op_mix_ratio=0.75,\n",
    "        )\n",
    "sc.tl.umap(adata2, a=2, b=1.4)\n",
    "sc.tl.leiden(adata2)\n",
    "for color in gene_ls:\n",
    "    if color == gene_dict[lineage]:\n",
    "        save = 'genes'\n",
    "    else:\n",
    "        save = color\n",
    "    sc.pl.umap(adata2, color = color, show = False,save=f'all_{color}.png')\n",
    "sc.tl.rank_genes_groups(adata2, f\"leiden\", method=\"wilcoxon\")\n",
    "sc.pl.rank_genes_groups_dotplot(\n",
    "    adata2,\n",
    "    groupby=f\"leiden\",\n",
    "    n_genes=int(150 / len(adata2.obs[f\"leiden\"].unique())),\n",
    "    show=True,\n",
    "    save = f'all.png'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a38fe3da",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ad30606",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
